gaphor/UML/classes/classconnect.py

```diff
@@ -152,24 +164,18 @@ def reconnect(self, handle, port):
         oend.subject.type = c.subject
         UML.model.set_navigability(line.subject, oend.subject, nav)
 
-    def disconnect_subject(self, handle):
-        """
-        Disconnect model element.
-        Disconnect property (memberEnd) too, in case of end of life for
-        Extension
+    def disconnect_subject(self, handle: Handle) -> None:
+        """Disconnect model element.
+
+        Disconnect property (memberEnd) in case of end of life for connection.
         """
         opposite = self.line.opposite(handle)
         c1 = self.get_connected(handle)
         c2 = self.get_connected(opposite)
-        if c1 and c2:
-            old: UML.Association = self.line.subject
-            del self.line.subject
```

Comment:
@amolenaar In order to keep the UML.Association including the navigability and aggregation, does this seem reasonable to you to not delete these on disconnect?

Reply:
~My original idea was to put that information on `AssociationItem.head_end` and `AssociationItem.tail_end`. These objects are created at construction time.~

~I missed that previously (was reviewing from my phone). Sorry 'bout that.~

EDIT: by looking at it for a second time I think this is much better, actually! An Association is also a Classifier, so it has a different role than, say, Generalization. Also, this will allow us to create Association Classes (or 3/4/5-way associations for that matter).

`len(self.line.subject.presentation) == 0`  could be rewritten as `not self.line.subject.presentation`.

Reply:
This will, however, disconnect the member end from everything. Including the model. So these `memberEnd` properties can not be reused... Same goes for `self.line.subject.unlink()`: the Association instance will be completely removed from the model. This causes errors when saving and loading the model.

Since the `Association` and its `memberEnd`s are created when the item is placed on the diagram, they should remain part of the model after disconnecting. Alternatively, we can just create the Association and add the memberEnds only when an association is connected. It will violate the multiplicity constraint, though (`[2..*]`).

I think this approach makes sense. An Association is also a Classifier, so it has a different role than, say, Generalization. Also, this will allow us to create Association Classes (or 3/4/5-way associations for that matter).
The trade-off is we give up the possibility to present an association in different diagrams. We can always work around this by offering such an option though the element editor, which may be fine, since I generally to not tend to use associations in multiple diagrams anyway.

Reply:
@amolenaar Thanks for the clarification on the `unlink`!

Ideally I think the associations should be reusable in many diagrams. A user should be able to grab a few elements in a model and then ask the model to display the relationships that already exist.

I thought we had to create the Association and memberEnds to display the navigability and aggregation. This is how it was a few commits ago, the association goes back to being a plain one as soon as you disconnect it.

Reply:
> Ideally I think the associations should be reusable in many diagrams. A user should be able to grab a few elements in a model and then ask the model to display the relationships that already exist.

That would be a nice feature. But it's different from:

1. create a diagram with 2 classes and draw an association between them (name it "foo")
2. create a new diagram, add the same 2 classes and draw another association
3. Will the association be:
   a. a new association
   b. be the already existing association named "foo"?

Currently we do option 3a. Which is fine IMHO.

> I thought we had to create the Association and memberEnds to display the navigability and aggregation. This is how it was a few commits ago, the association goes back to being a plain one as soon as you disconnect it.

Indeed you'll need both Association and the Properties at the memberEnds. Since the logic on navigability lies at the model. Not in the diagram items. It should, however, still disconnect the "type"s at the end. In the connect() method, the association ends are paired with the classes they connect to. On disconnect, this relation should be removed I think. Otherwise the model and the diagrams are no longer in sync.

Reply:
Thanks good catch. I don't think I was properly connecting the the type to the element either. How does this look?